File: Week_06/house_robber_ii/Solution.java
package house_robber_ii;

/**
 * 动态规划
 * 把环状排列的房子简化为两个单排列的
 * 1. 偷第一个，不偷最后一个
 * 2. 偷最后一个，不偷第一个
 * 最大金额 = 两种情况的较大值
 */
class Solution {
    public int rob(int[] nums) {
        int n = nums != null ? nums.length : 0;
        if (n < 2) {
            return n == 0 ? 0 : nums[0];
        }
        return Math.max(rob(nums, 0, n - 1), rob(nums, 1, n));
    }
    public int rob(int[] nums, int start, int end) {
        int prev = 0, curr = 0;
        for (int i = start; i < end; i++) {
            int next = Math.max(curr, prev + nums[i]);
            prev = curr;
            curr = next;
        }
        return curr;
    }
}

File: Week_06/house_robber_ii/Solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob_helper(start, end):
            pre = cur = 0
            for i in range(start, end):
                cur, pre = max(cur, pre + nums[i]), cur
            return cur

        n = len(nums)
        return max(rob_helper(0, n - 1), rob_helper(1, n)) if n > 1 else nums[0] if n == 1 else 0

